### PR TITLE
fix: inheritance for serverOnly scoped classes and fields

### PR DESCRIPTION
**Pull request by BenAuerDev**
This PR tries to adjust the `SerializableModelLibraryGenerator` to handle inherited serverOnly fields correctly. As the parentClass declares them as `Object?` but the child-class parses their real type (`int`, `String` etc.) we now check in the methods building `_buildCopyWithMethod` and `_buildAbstractCopyWithMethod` and pass them to the childs copyWith method as expected

Furthermore a test case was added to `json_protocol_test.dart` making sure `serverOnly` fields are not leaked when inherited. 

Another test was added to `server_code_generator/inheritance_class_test.dart` checking if nullable fields are typed as expected in the `copyWith` method. 

And a `restriction` and `test` was added to only allow extending a `serverOnly` class if the child is marked as `serverOnly` aswell. 

Selection_003

linked to: #1873 and #1618 

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [x] Any breaking changes are documented below.

## Breaking changes

no breaking changes

### CODE REVIEW

File: tests/serverpod_test_server/lib/src/generated/scopes/scope_server_only_field_child.dart [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+  String childFoo;
+
+  @override
+  ScopeServerOnlyFieldChild copyWith({
+    Object? allScope,
+    Object? serverOnlyScope,
+    Object? nested,
+    String? childFoo,
+  });
```
Comment (SandPod) [inline tests/serverpod_test_server/lib/src/generated/scopes/scope_server_only_field_child.dart:58]: Question: Shouldn't a typed interface be used here? Isn't it only inherited classes that do not have support for type copy with interfaces?

File: tools/serverpod_cli/lib/src/analyzer/models/validation/restrictions.dart [hunk 2]
```diff
@@ -308,6 +308,15 @@ class Restrictions {
       ];
     }
 
+    if (parentClass.serverOnly && !documentDefinition!.serverOnly) {
```
[the remark this answers is not in the document]
Comment (SandPod) [reply tools/serverpod_cli/lib/src/analyzer/models/validation/restrictions.dart:311]: Please add a test for this as well.

File: tests/serverpod_test_server/lib/src/endpoints/server_only_scoped_field_model_child.dart [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import 'dart:async';
+
+import 'package:serverpod/serverpod.dart';
+import 'package:serverpod_test_server/src/generated/protocol.dart';
+
+class ServerOnlyScopedFieldChildModelEndpoint extends Endpoint {
+  Future<ScopeServerOnlyField> getProtocolField(
```
Comment (SandPod) [inline tests/serverpod_test_server/lib/src/endpoints/server_only_scoped_field_model_child.dart:7]: Suggestion: This endpoint doesn't return `ScopeServerOnlyFieldChild` which is what we want to test.

Please change the return type to catch this specific case.

**Review by SandPod — COMMENTED** [review]
Nice work!

A couple of suggestions for improvements and a couple of questions around the work.

It seems like we might have broken the typed copy with interface a bit.

File: tests/serverpod_test_server/lib/src/endpoints/server_only_scoped_field_model_child.dart [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+  Future<ScopeServerOnlyFieldChild> getProtocolField(
+    Session session,
+  ) async {
+    return ScopeServerOnlyFieldChild(
+        serverOnlyScope: Types(anInt: 2),
+        nested: ScopeServerOnlyField(
+          allScope: Types(anInt: 1),
+          serverOnlyScope: Types(anInt: 2),
+        ),
+        childFoo: 'childFoo');
```
Comment (SandPod) [inline tests/serverpod_test_server/lib/src/endpoints/server_only_scoped_field_model_child.dart:16]: Suggestion(NIT): Add comma for better formatting here.

File: tools/serverpod_cli/lib/src/analyzer/models/validation/restrictions.dart [hunk 5]
```diff
@@ -308,6 +308,21 @@ class Restrictions {
...
+    var currentModel =
+        parsedModels.findByClassName(documentDefinition!.className);
+
+    var ancestorServerOnlyClass =
+        _findServerOnlyClassInParentClasses(currentModel);
+
+    if (!documentDefinition!.serverOnly && ancestorServerOnlyClass != null) {
+      return [
+        SourceSpanSeverityException(
+          'The class "${documentDefinition!.className}" cannot inherit from the server-only class "${ancestorServerOnlyClass.className}" unless it is also marked as "serverOnly".',
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/analyzer/models/validation/restrictions.dart:320]: Suggestion: How about `Cannot extend a "serverOnly" class in the inheritance chain ("${ancestorServerOnlyClass.className}") unless class is marked as "serverOnly".`?

File: tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/inheritance_test.dart [hunk 6]
```diff
@@ -213,4 +213,90 @@ void main() {
...
+
+    var error = collector.errors.first;
+    expect(
+      error.message,
+      'The class "ExampleChildClass" cannot inherit from the server-only class "Example" unless it is also marked as "serverOnly".',
+    );
+  });
+
+  test(
+      'Given a serverOnly child-class, When the parent-class not serverOnly but the grandparent-class is, then error is collected that a client class cannot extend a serverOnly class',
```
Comment (SandPod) [inline tools/serverpod_cli/test/analyzer/models/stateful_analyzer/model_validation/inheritance_test.dart:257]: Suggestion(NIT):
```suggestion
      'Given a serverOnly child-class, When the parent-class is not serverOnly but the grandparent-class is, then error is collected that a client class cannot extend a serverOnly class',
```

File: tools/serverpod_cli/lib/src/generator/dart/library_generators/model_library_generator.dart [hunk 7]
```diff
@@ -1138,19 +1139,23 @@ class SerializableModelLibraryGenerator {
...
         subDirParts: classDefinition.subDirParts,
         config: config,
       );
 
+      var isInheritedField = classDefinition.inheritedFields.contains(field);
+
+      var type = field.type.nullable && isInheritedField
+          ? refer('Object?')
+          : fieldType;
+
```
Comment (SandPod) [inline tools/serverpod_cli/lib/src/generator/dart/library_generators/model_library_generator.dart:1154]: Suggestion: Could we add some test for this change?

**Review by SandPod — COMMENTED** [review]
almost the same result but with less changes. I like that! A couple of more comments and suggestions.

File: tools/serverpod_cli/lib/src/generator/dart/library_generators/model_library_generator.dart (hunk 7 above)
Comment (BenAuerDev) [reply tools/serverpod_cli/lib/src/generator/dart/library_generators/model_library_generator.dart:1154]: Yes I'll add one in `server_code_generator > inheritance_class_test.dart` checking if 
inherited-nullable => `Object?` 
local-nullable => typed (int?, String?, ...)

File: tools/serverpod_cli/test/generator/dart/server_code_generator/inheritance_class_test.dart [hunk 8]
```diff
@@ -348,4 +348,68 @@ void main() {
...
+        );
+
+        test(
+            'with named params where nullable-inherited fields are "Object?" and nullable-local fields are typed',
+            () {
+          expect(
+            copyWithMethod?.parameters?.toSource(),
+            '({int? id, Object? name, int? age})',
+          );
+        }, skip: copyWithMethod == null);
```
Comment (SandPod) [inline tools/serverpod_cli/test/generator/dart/server_code_generator/inheritance_class_test.dart:411]: Suggestion: Please split these two assertions into two different tests. One that validated the inherited nullable field and one that validatesthe nullable local field.

File: tools/serverpod_cli/test/generator/dart/server_code_generator/inheritance_class_test.dart [hunk 9]
```diff
@@ -348,4 +348,68 @@ void main() {
...
+
+    var codeMap = generator.generateSerializableModelsCode(
+      models: models,
+      config: config,
+    );
+
+    var compilationUnit =
+        parseString(content: codeMap[childExpectedFilePath]!).unit;
+
+    group('Then the $childClassName', () {
```
Comment (SandPod) [inline tools/serverpod_cli/test/generator/dart/server_code_generator/inheritance_class_test.dart:392]: Suggestion(NIT): change `Then` to `then`.

**Review by SandPod — APPROVED** [review]
LGTM with the last modifications that I commented on!

Nice work!

**Pull request merged**